extensions-contrib/scan-query/src/main/java/io/druid/query/scan/ScanQueryEngine.java

```diff
@@ -0,0 +1,242 @@
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeoutException;
+
+public class ScanQueryEngine
+{
+  private static final SelectQueryEngine.SelectStrategyFactory STRATEGY_FACTORY = new SelectQueryEngine.SelectStrategyFactory();
+  public Sequence<ScanResultValue> process(
+      final ScanQuery query,
+      final Segment segment,
+      final Map<String, Object> responseContext
+  )
+  {
+    if (responseContext.get("count") != null) {
+      int count = (int) responseContext.get("count");
+      if (count >= query.getLimit()) {
+        return Sequences.empty();
+      }
+    }
+    final Long timeoutAt = (long) responseContext.get("timeoutAt");
+    final long start = System.currentTimeMillis();
+    final StorageAdapter adapter = segment.asStorageAdapter();
+
+    if (adapter == null) {
+      throw new ISE(
+          "Null storage adapter found. Probably trying to issue a query against a segment being memory unmapped."
+      );
+    }
+
+    // at the point where this code is called, only one datasource should exist.
+    String dataSource = Iterables.getOnlyElement(query.getDataSource().getNames());
+
+    List<String> allDims = Lists.newLinkedList(adapter.getAvailableDimensions());
+    List<String> allMetrics = Lists.newLinkedList(adapter.getAvailableMetrics());
+    final List<String> allColumns = Lists.newLinkedList();
+    if (query.getColumns() != null && !query.getColumns().isEmpty()) {
+      if (!query.getColumns().contains(ScanResultValue.timestampKey)) {
+        allColumns.add(ScanResultValue.timestampKey);
+      }
+      allColumns.addAll(query.getColumns());
+      allDims.retainAll(query.getColumns());
+      allMetrics.retainAll(query.getColumns());
+    }
+    if (query.getColumns() == null || query.getColumns().isEmpty()) {
+      if (!allDims.contains(ScanResultValue.timestampKey)) {
+        allColumns.add(ScanResultValue.timestampKey);
+      }
+      allColumns.addAll(allDims);
+      allColumns.addAll(allMetrics);
+    }
+    final List<DimensionSpec> dims = DefaultDimensionSpec.toSpec(allDims);
+    final List<String> metrics = allMetrics;
+
+    final List<Interval> intervals = query.getQuerySegmentSpec().getIntervals();
+    Preconditions.checkArgument(intervals.size() == 1, "Can only handle a single interval, got[%s]", intervals);
+
+    final String segmentId = segment.getIdentifier();
+
+    final Filter filter = Filters.convertToCNFFromQueryContext(query, Filters.toFilter(query.getDimensionsFilter()));
+
+    if (responseContext.get("count") == null) {
+      responseContext.put("count", 0);
+    }
+    final int limit = query.getLimit() - (int) responseContext.get("count");
+    return Sequences.concat(
+        Sequences.map(
+            adapter.makeCursors(
+                filter,
+                intervals.get(0),
+                VirtualColumns.EMPTY,
+                QueryGranularities.ALL,
```

Comment:
It should be ALL. I only want 1 cursor, and can read all data in one iterator.
If it is NONE, it will create 1 cursor per millisecond, see CursorSequenceBuilder `Iterable<Long> iterable = gran.iterable(interval.getStartMillis(), interval.getEndMillis());`
and BaseQueryGranularity.iterable() and AllGranularity.iterable() returns single element iterable